File: src/BRCurtidas.FSharp.Benchmarks/DomainBenchmarks.fs
namespace BRCurtidas.FSharp.Benchmarks

open BRCurtidas.FSharp.Data
open BenchmarkDotNet.Attributes
open Chessie.ErrorHandling
open BenchmarkDotNet.Attributes.Jobs
open BenchmarkDotNet.Engines

[<MemoryDiagnoser>]
[<SimpleJob(RunStrategy.ColdStart, launchCount = 1, targetCount = 1)>]
type DomainBenchmarks() =
    let mutable orders : Entity<Order> list = []

    let errorMsg errors = sprintf "Error. Generated entity must not result in a failure, but resulted in: %A" errors

    let generateOrders =
        let rec loop current orders max =
            if current > max then
                orders
            else
                let entity = trial {
                    let! account = Account.create "ivelten" "password"
                    let! address = Address.create "Rua M. Francisco de Moura, 251" (Some "CS4")
                    let! user = User.create "Ismael" "ismaelcarlosvelten@gmail.com" "105.643.607-75" "21980836100" account address
                    let! product = Product.create 21.5M true "Pacote com curtidas nacionais no Facebook" "Curtidas Facebook Nacionais"
                    let payment = Some { Reference = "12345F"; Method = PaymentMethod.PayPal }
                    let! network = SocialNetworkAccount.create SocialNetwork.Facebook "https://www.facebook.com/ismaelcarlosvelten"
                    let! order = Order.create user product payment OrderStatus.WaitingPayment 200 network
                    return! Entity.create current order }
                match entity with
                | Ok (entity, _) -> loop (current + 1L) (entity :: orders) max
                | Bad errors -> failwith (errorMsg errors)
        loop 1L []

    let setAccount (entity : Entity<Order>) = trial {
        let! account = Account.create "aanjos" "password2"
        let! user = User.create "André" "aanjos@gmail.com" "158.844.757-05" "21965448522" account entity.Value.User.Address
        return { entity with Value = { entity.Value with User = user } } }

    let changeOrders orders =
        orders
        |> List.map (fun x -> setAccount x)
        |> List.map (fun x ->
            match x with
            | Ok (entity, _) -> entity
            | Bad errors -> failwith (errorMsg errors))

    [<Benchmark>]
    member __.GenerateOneHundredOrders() =
        orders <- generateOrders 100L

    [<Benchmark>]
    member __.GenerateOneThousandOrders() =
        orders <- generateOrders 1000L

    [<GlobalSetup(Target = "ChangeOneHundredOrders")>]
    member __.SetupChangeOneHundredOrders() =
        orders <- generateOrders 100L

    [<Benchmark>]
    member __.ChangeOneHundredOrders() =
        orders <- changeOrders orders

    [<GlobalSetup(Target = "ChangeOneThousandOrders")>]
    member __.SetupChangeChangeOneThousandOrders() =
        orders <- generateOrders 1000L

    [<Benchmark>]
    member __.ChangeOneThousandOrders() =
        orders <- changeOrders orders

File: src/BRCurtidas.CSharp.Benchmarks/DomainBenchmarks.cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Attributes.Jobs;
using BenchmarkDotNet.Engines;
using BRCurtidas.CSharp.Common.DomainTypes;
using BRCurtidas.CSharp.Data;
using System;
using System.Collections.Generic;

namespace BRCurtidas.CSharp.Benchmarks
{
    [MemoryDiagnoser]
    [SimpleJob(RunStrategy.ColdStart, launchCount: 1, targetCount: 1)]
    public class DomainBenchmarks
    {
        private readonly List<Entity<Order>> orders = new List<Entity<Order>>();

        private void GenerateOrders(List<Entity<Order>> orders, long max)
        {
            for (long i = 1; i <= max; i++)
            {
                var product = new Product
                {
                    Price = new Currency(21.5M),
                    Description = new String100("Pacote com curtidas nacionais no Facebook"),
                    Title = new String50("Curtidas Facebook Nacionais"),
                    Enabled = true
                };

                var account = new Account
                {
                    UserName = new UserName("ivelten"),
                    PasswordHash = PasswordHash.FromPassword("password")
                };

                var address = new Address
                {
                    Line1 = new String100("Rua M. Francisco de Moura, 251"),
                    Line2 = new String100("CS4")
                };

                var user = new User
                {
                    Account = account,
                    Address = address,
                    Cpf = new Cpf("105.643.607-75"),
                    Email = new EmailAddress("ismaelcarlosvelten@gmail.com"),
                    Name = new String50("Ismael"),
                    Phone = new PhoneNumber("21980836100")
                };

                var payment = new Payment
                {
                    Reference = "12345F",
                    Method = PaymentMethod.PayPal
                };

                var socialNetworkAccount = new SocialNetworkAccount
                {
                    Profile = new Url("https://www.facebook.com/ismaelcarlosvelten"),
                    Type = SocialNetworkType.Facebook
                };

                var qty = new Quantity(200);
                var price = new Currency(qty.Value * product.Price.Value);

                var order = new Order
                {
                    User = user,
                    Payment = payment,
                    Product = product,
                    Quantity = qty,
                    Price = price,
                    SocialNetworkAccount = socialNetworkAccount,
                    Status = OrderStatus.WaitingPayment,
                    Created = DateTime.Now
                };

                var entity = new Entity<Order>
                {
                    Id = new Id(i),
                    Value = order
                };

                orders.Add(entity);
            }
        }

        private void ChangeOrders(List<Entity<Order>> orders)
        {
            foreach (var order in orders)
            {
                order.Value.User.Name = new String50("André");
                order.Value.User.Email = new EmailAddress("aanjos@gmail.com");
                order.Value.User.Cpf = new Cpf("158.844.757-05");
                order.Value.User.Phone = new PhoneNumber("21965448522");
                order.Value.User.Account.UserName = new UserName("aanjos");
                order.Value.User.Account.PasswordHash = PasswordHash.FromPassword("password2");
            }
        }

        [Benchmark]
        public void GenerateOneHundredOrders()
        {
            GenerateOrders(orders, 100L);
        }

        [Benchmark]
        public void GenerateOneThousandOrders()
        {
            GenerateOrders(orders, 1000L);
        }

        [GlobalSetup(Target = nameof(ChangeOneHundredOrders))]
        public void SetupChangeOneHundredOrders()
        {
            GenerateOrders(orders, 100L);
        }

        [Benchmark]
        public void ChangeOneHundredOrders()
        {
            ChangeOrders(orders);
        }

        [GlobalSetup(Target = nameof(ChangeOneThousandOrders))]
        public void SetupChangeOneThousandOrders()
        {
            GenerateOrders(orders, 1000L);
        }

        [Benchmark]
        public void ChangeOneThousandOrders()
        {
            ChangeOrders(orders);
        }
    }
}